src/connector/src/sink/opensearch.rs

```diff
@@ -0,0 +1,150 @@
+use serde_json::Value;
+use tonic::async_trait;
+
+use super::catalog::desc::SinkDesc;
+use super::elasticsearch_opensearch_common::{
+    validate_config, ElasticSearchOpenSearchConfig, ElasticSearchOpenSearchFormatter,
+};
+use super::log_store::DeliveryFutureManagerAddFuture;
+use super::writer::{
+    AsyncTruncateLogSinkerOf, AsyncTruncateSinkWriter, AsyncTruncateSinkWriterExt,
+};
+use super::{DummySinkCommitCoordinator, Sink, SinkError, SinkParam, SinkWriterParam};
+use crate::sink::Result;
+
+pub const OPENSEARCH_SINK: &str = "opensearch_rust";
+
+pub type OpenSearchSinkDeliveryFuture =
+    impl TryFuture<Ok = (), Error = SinkError> + Unpin + 'static;
+
+#[derive(Debug)]
+pub struct OpenSearchSink {
+    config: ElasticSearchOpenSearchConfig,
+    schema: Schema,
+    pk_indices: Vec<usize>,
+}
+
+#[async_trait]
+impl TryFrom<SinkParam> for OpenSearchSink {
+    type Error = SinkError;
+
+    fn try_from(param: SinkParam) -> std::result::Result<Self, Self::Error> {
+        let schema = param.schema();
+        let config = ElasticSearchOpenSearchConfig::from_btreemap(param.properties, &schema)?;
+        Ok(Self {
+            config,
+            schema,
+            pk_indices: param.downstream_pk,
+        })
+    }
+}
+
+impl Sink for OpenSearchSink {
+    type Coordinator = DummySinkCommitCoordinator;
+    type LogSinker = AsyncTruncateLogSinkerOf<OpenSearchSinkWriter>;
+
+    const SINK_NAME: &'static str = OPENSEARCH_SINK;
+
+    fn is_sink_decouple(_desc: &SinkDesc, user_specified: &SinkDecouple) -> Result<bool> {
+        match user_specified {
+            SinkDecouple::Default | SinkDecouple::Enable => Ok(true),
+            SinkDecouple::Disable => Ok(false),
+        }
+    }
+
+    async fn validate(&self) -> Result<()> {
+        validate_config(&self.config, &self.schema)?;
+        let client = self.config.build_opensearch_client()?;
+        client.ping().send().await?;
+        Ok(())
+    }
+
+    async fn new_log_sinker(&self, _writer_param: SinkWriterParam) -> Result<Self::LogSinker> {
+        Ok(OpenSearchSinkWriter::new(
+            self.config.clone(),
+            self.schema.clone(),
+            self.pk_indices.clone(),
+        )?
+        .into_log_sinker(usize::MAX))
+    }
+}
+
+pub struct OpenSearchSinkWriter {
+    client: Arc<OpenSearch>,
+    formatter: ElasticSearchOpenSearchFormatter,
+}
+
+impl OpenSearchSinkWriter {
+    pub fn new(
+        config: ElasticSearchOpenSearchConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+    ) -> Result<Self> {
+        let client = Arc::new(config.build_opensearch_client()?);
+        let formatter = ElasticSearchOpenSearchFormatter::new(
+            pk_indices,
+            &schema,
+            config.delimiter,
+            config.index_column,
+            config.index,
+        )?;
+        Ok(Self { client, formatter })
+    }
+}
+
+impl AsyncTruncateSinkWriter for OpenSearchSinkWriter {
```

Comment:
The implementation of the two sink writers are basically the same. Can we introduce a new trait to extract the common logic of the two sinks and implement only a generic sink writer shared by the two sinks?

Reply:
It seems to be a bit difficult, because here it looks like the name is the same, but the package introduced is two packages, the